swarm_dial.go

```diff
@@ -343,23 +343,87 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	if len(goodAddrs) == 0 {
 		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
 	}
-	goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
+
+	///////// Partition and rank addresses
+	// Partition into Fd and non-Fd consuming addresses as we first we want
+	// to try dialling ALL Non-Fd consuming addresses.
 	nonBackoff := false
+	var fdAddrs []ma.Multiaddr
+	var nonFdAddrs []ma.Multiaddr
+
 	for _, a := range goodAddrs {
 		// skip addresses in back-off
 		if !s.backf.Backoff(p, a) {
 			nonBackoff = true
-			goodAddrsChan <- a
+			if s.IsFdConsumingAddr(a) {
+				fdAddrs = append(fdAddrs, a)
+			} else {
+				nonFdAddrs = append(nonFdAddrs, a)
+			}
 		}
 	}
-	close(goodAddrsChan)
 	if !nonBackoff {
 		return nil, ErrDialBackoff
 	}
+
+	/*
+		This slice-to-chan code is temporary, the peerstore can currently provide
+		a channel as an interface for receiving addresses, but more thought
+		needs to be put into the execution. For now, this allows us to use
+		the improved rate limiter, while maintaining the outward behaviour
+		that we previously had (halting a dial when we run out of addrs)
+	*/
+	// addrsToChanFnc transforms a slice of addresses to a closed channel of addresses.
+	addrsToChanFnc := func(goodAddrs []ma.Multiaddr) chan ma.Multiaddr {
+		goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
+		for _, a := range goodAddrs {
+			goodAddrsChan <- a
+		}
+		close(goodAddrsChan)
+		return goodAddrsChan
+	}
+
+	// ranks addresses in descending order of preference for dialing
+	// local addrs > non-relay public addrs > relay-addrs
+	rankAddrsFnc := func(addrs []ma.Multiaddr) []ma.Multiaddr {
+		var localAddrs []ma.Multiaddr
+		var relayAddrs []ma.Multiaddr
+		var others []ma.Multiaddr
+
+		for _, a := range addrs {
+			if manet.IsPrivateAddr(a) {
+				localAddrs = append(localAddrs, a)
+			} else if _, err := a.ValueForProtocol(ma.P_CIRCUIT); err == nil {
+				relayAddrs = append(relayAddrs, a)
+			} else {
+				others = append(others, a)
+			}
+		}
+
+		return append(append(localAddrs, others...), relayAddrs...)
+	}
+
+	// Sort/Rank by preference
+	nonFdAddrs = rankAddrsFnc(nonFdAddrs)
+	fdAddrs = rankAddrsFnc(fdAddrs)
+
 	/////////
+	// first try to get a connection to a non-fd consuming addr and if that fails,
+	// try to get a connection to a fd consuming addr.
+	var connC transport.CapableConn
+	var errNonFd *DialError
+	var dialErr *DialError
+
+	connC, errNonFd = s.dialAddrs(ctx, p, addrsToChanFnc(nonFdAddrs))
```

Comment:
We _can_ just get rid of the channel on this interface to simplify it.

Reply:
Agreed. This is done.